ignite/metrics/nlp/bleu.py

```diff
@@ -122,14 +126,105 @@ def __init__(
         smooth: str = "no_smooth",
         output_transform: Callable = lambda x: x,
         device: Union[str, torch.device] = torch.device("cpu"),
+        average: str = "macro",
     ):
         if ngram <= 0:
             raise ValueError(f"ngram order must be greater than zero (got: {ngram})")
         self.ngrams_order = ngram
         self.weights = [1 / self.ngrams_order] * self.ngrams_order
         self.smoother = _Smoother(method=smooth)
+
+        if average not in ["macro", "micro"]:
+            raise ValueError(f'Average must be either "macro" or "micro" (got: {average})')
+        self.average = average
+        self.p_numerators: Counter = Counter()
+        self.p_denominators: Counter = Counter()
+        self.hyp_length_sum = 0
+        self.ref_length_sum = 0
+
         super(Bleu, self).__init__(output_transform=output_transform, device=device)
 
+    def _n_gram_counter(
+        self, references: Sequence[Sequence[Sequence[Any]]], candidates: Sequence[Sequence[Any]],
+    ) -> None:
+        if len(references) != len(candidates):
+            raise ValueError(
+                f"nb of candidates should be equal to nb of reference lists ({len(candidates)} != "
+                f"{len(references)})"
+            )
+
+        # Iterate through each hypothesis and their corresponding references.
+        for refs, hyp in zip(references, candidates):
+            # For each order of ngram, calculate the numerator and
+            # denominator for the corpus-level modified precision.
+            for i in range(1, self.ngrams_order + 1):
+                numerator, denominator = modified_precision(refs, hyp, i)
+                self.p_numerators[i] += numerator
+                self.p_denominators[i] += denominator
+
+            # Calculate the hypothesis lengths
+            self.hyp_length_sum += len(hyp)
+
+            # Calculate the closest reference lengths.
+            self.ref_length_sum += _closest_ref_length(refs, len(hyp))
+
+    def _brevity_penalty_smoothing_macro(self) -> float:
```

Comment:
Only one version of `_brevity_penalty_smoothing` must be provided (code duplication antipattern) and it should be static as suggested above.